subsys/net/ip/utils.c

```diff
@@ -486,32 +486,48 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
 #include <syscalls/net_addr_pton_mrsh.c>
 #endif /* CONFIG_USERSPACE */
 
-static uint16_t calc_chksum(uint16_t sum, const uint8_t *data, size_t len)
+/* Word based checksum calculation based on:
+ * https://blogs.igalia.com/dpino/2018/06/14/fast-checksum-computation/
+ * It’s not necessary to add octets as 16-bit words. Due to the associative property of addition,
+ * it is possible to do parallel addition using larger word sizes such as 32-bit or 64-bit words.
+ * In those cases the variable that stores the accumulative sum has to be bigger too.
+ * Once the sum is computed a final step folds the sum to a 16-bit word (adding carry if any).
+ */
+uint16_t calc_chksum(uint16_t sum_in, const uint8_t *data, size_t len)
 {
-	const uint8_t *end;
-	uint16_t tmp;
-
-	end = data + len - 1;
+	uint64_t sum = (uint64_t)ntohs(sum_in);
+	uint32_t *p = (uint32_t *) data;
+	uint16_t i = 0;
+	size_t pending = len;
```

Comment:
Would the following make sense?
```suggestion
	uint32_t *p;
	uint16_t i = 0;
	size_t pending = len;

	while ((((uintptr_t)data & 0x03) != 0) && (pending > 0)) {
		sum += *data;
		pending--;
		data++;
	}

	p = (uint32_t *)data;
```

That way we guarantee that `data` is aligned properly before actually using the address with `uint32_t` pointer.

Reply:
Thanks for the good hint, I took it as basis, but needed some fiddling to get the byte alignment right